RFCs/0004-add-pdk-publish.md

```diff
@@ -0,0 +1,123 @@
+      - If given tarball cannot be found, exits with an error.
+
+
+    - `--package-dir=<path>`
+      - Looks in current module’s `metadata.json` for namespace, module name, and version and then attempts to publish `<namespace>-<module_name>-<version>.tar.gz` from given path.
+      - If file matching the `<namespace>-<module_name>-<version>.tar.gz` cannot be found in the given path, exits with an error.
+      
+
+    - `--package-dir=<path> <path to tarball>`
+      - Logs a warning that `--package-dir` option is being ignored in favor of pathed tarball argument.
+      - Attempts to publish given pathed tarball.
+      - If given pathed tarball cannot be found, exits with an error.
+
+
+    - `(any valid options/args) --force`
+      - Attempts to publish without stopping to ask for confirmation.
+      - Note that command will still fail, regardless of the `--force` flag, if a release of the module with the same version already exists on the Forge.
+
+- Defaults to asking for confirmation when run interactively, but this can be disabled. (See `--force` flag.)
+
+- When run non-interactively (e.g. as part of a continuous delivery pipeline), it will default to not asking for confirmation (`--force` behavior) and exit non-zero if it would require any user input to complete successfully.
+
+- Need to track namespace/api-key mapping in user-level config.
+    - Look inside tarball to determine namespace
+    - Check to see if config contains an api-key for that namespace
+    - Prompt for api-key if not (can use existing wrappers around tty-prompt)
+
+- Need to be able to provide api-key as an environment variable with the name PDK_FORGE_KEY
+
+### Examples
+
+#### Publishing a tarball matching data from metadata.json when no options / arguments are provided 
+```
+$ pdk publish  
+pdk (INFO): Attempting to publish puppetlabs-mysql-0.1.0.tar.gz  
+
+Once published to the Forge, your module will be made available for public download. Continue? (y/N) yes  
+pdk (INFO): Authenticating to the Forge API using key with namespace: puppetlabs  
+Publish successful, new release available at https://forge.puppet.com/v3/releases/puppetlabs-mysql-0.1.0  
+```
+
+#### Attempting to publish a release that already exists on the Forge, bypassing confirmation with `--force`
+```
+$ pdk publish --force  
+pdk (INFO): Attempting to publish puppetlabs-mysql-0.1.0.tar.gz  
+pdk (INFO): Authenticating to the Forge API using key with namespace: puppetlabs  
+Error: 409 Conflict, a release version 0.1.0 for module puppetlabs-mysql already exists on the Forge. To create a new version of your module to publish, update the version number in metadata.json and run `pdk build`.
+```
+
+#### Prompt for API key
+```
+$ pdk publish  
+pdk (INFO): Attempting to publish puppetlabs-mysql-0.1.1.tar.gz  
+
+No API key was found for namespace: puppetlabs in PDK user config or as an environment variable. If you have an API key you'd like to use, enter it now:
+--> **************  
```

Comment:
Maybe the API key prompt should come after the confirmation prompt?

Reply:
That seems reasonable. Do you think the confirmation should actually be before the `pdk (INFO): Attempting to publish` as well?

Reply:
I think it's useful for the user to see what module and version is about to be published before we ask for confirmation, that way they can see (and confirm) that it's doing what they expected.

We can workshop exactly what the message says during implementation.

Reply:
I'm not sure why this isn't tagged as "outdated", but I've made that change.